### PR TITLE
feat(localchain): control accounts on and make queries to Agoric's Cosmos side

### LINKED ISSUE #8926: orchestration: create vat-localchain
## What is the Problem Being Solved?

In an effort to break down #8624 and isolate some interwoven dependencies, expose `makeLocalChain`, a function that allows consumers to create chain accounts and use them to send queries and transactions.

### PR DESCRIPTION
**Pull request by michaelfig**
closes: #8926
refs: #8624, #8895

## Description

This is an end-to-end of a simplistic Agoric localchain API to expose transactions and queries that an orchestration JS contract could use to control a local `agoric1...` account and make state queries.

It's split off from a similar feature in #8624 to make the reviews possible in smaller pieces.

### Security Considerations

tbd

### Scaling Considerations

tbd

### Documentation Considerations

This is a pragmatic step towards the Orchestration API.  I expect it will be documented after the Orchestration API is settled and we rework localchain.js to conform to it.

### Testing Considerations

There are unit tests of the Golang `x/vlocalchain` module, and an `agoric-sdk/a3p-integration/proposals/b:localchain` test that verifies the localchain proposal builder and runs a simple on-chain test after the proposal passes.

An upgrade test of `vat-localchain.js` is necessary.

I was also able to test manually in the REPL and get the results I was expecting:

image

### Upgrade Considerations

The new localchain objects are durable-ready via Zones.

### CODE REVIEW

File: a3p-integration/proposals/b:localchain/package.json
Comment (turadg) [inline a3p-integration/proposals/b:localchain/package.json]:
```suggestion
    "@agoric/synthetic-chain": "^0.0.7",
```

File: a3p-integration/proposals/b:localchain/post.test.js
Comment (turadg) [inline a3p-integration/proposals/b:localchain/post.test.js]: why is this a separate proposal from `a:upgrade-next`?

File: a3p-integration/proposals/g:restart-vats/README.md
Comment (turadg) [inline a3p-integration/proposals/g:restart-vats/README.md:1]: I'm having second thoughts about the value of this proposal. I'd be okay dropping it. Meanwhile, `g` is a fine way to make room

File: golang/cosmos/app/app.go
Comment (turadg) [inline golang/cosmos/app/app.go]: please provide more explanation in the comment for why these are disabled here

File: packages/agoric-cli/src/commands/wallet.js
Comment (turadg) [inline packages/agoric-cli/src/commands/wallet.js:11]: what motivated this?

File: packages/builders/scripts/vats/init-localchain.js
Comment (turadg) [inline packages/builders/scripts/vats/init-localchain.js]: nifty, but a little magic to have it in the `init` proposal. why not put the test concern into another proposal?

File: packages/cosmic-swingset/src/chain-main.js
Comment (turadg) [inline packages/cosmic-swingset/src/chain-main.js]: we should be consistent with the casing. if it's supposed to be ALL_CAPS then let's change it in the `internal` package. But I don't think it should be. `BridgeId` is a namespace and the values are ALL_CAPS. 

i.e. `BridgeId.PROVISION_SMART_WALLET`, which is much easier to read than `BRIDGE_ID.PROVISION_SMART_WALLET`

so I request,
```suggestion
  BridgeId,
  BridgeKeys,
```

If this isn't agreeable please point to some documentation for the style in the PR currently

File: packages/cosmic-swingset/src/chain-main.js
Comment (turadg) [inline packages/cosmic-swingset/src/chain-main.js]: minor readability suggestion
```suggestion
        const portKey = Object.keys(BRIDGE_KEYS).find(
          k => BRIDGE_KEYS[key] === dstID,
        ) || [`${dstID}${PORT_SUFFIX}`];
```

modulo other requests

File: packages/deployment/Makefile
Comment (turadg) [inline packages/deployment/Makefile]: what motivates this?

it adds a level of indirection for knowing what's happening and one can't tell in this file what scripts might overriding. I don't see any uses so I assume it's for CLI dev usage but is it necessary in SCM?

File: packages/internal/src/config.js
Comment (turadg) [inline packages/internal/src/config.js]: this name confuses me because a lookup usually describes what you get, and you're getting `BridgeId` values same as the object above.

Would one of these make more sense?
- `BridgeIdByPort`
- `BridgePortAlias`

File: packages/vats/src/localchain.js
Comment (turadg) [inline packages/vats/src/localchain.js]: is each message really `M.any()`?

File: packages/vats/src/localchain.js
Comment (turadg) [inline packages/vats/src/localchain.js]: do you mean "example"? Or perhaps "subset" of the queries that need to be included?

File: packages/vats/src/localchain.js
Comment (turadg) [inline packages/vats/src/localchain.js:24]: please annotate signature so that the method returns aren't `any`

```suggestion
    /**
     * @param {import('./types.js').ScopedBridgeManager} system
     * @param {string} address
     * @param {{ query: (messages: any[]) => Promise<any> }} chain
     */
    (system, address, chain) => ({ system, address, chain }),
```

File: packages/vats/src/localchain.js
Comment (turadg) [inline packages/vats/src/localchain.js]: please annotate
```suggestion
  zone.exoClass('LocalChain', LocalChainI,
    /** @param {import('./types.js').ScopedBridgeManager} system */
    system => ({ system }),
```

File: packages/vats/src/localchain.js
Comment (turadg) [inline packages/vats/src/localchain.js:30]: can we expect to expand this with any sort of `chain.query`? I wonder if we can have some consistency with mapping to the `@type` namespace.

File: packages/vats/src/proposals/localchain-proposal.js
Comment (turadg) [inline packages/vats/src/proposals/localchain-proposal.js]: this dynamic manifest makes auditing harder. apropos the question above about keeping the testing concern out of `init-localchain`.

File: packages/cosmic-swingset/src/chain-main.js
Comment (turadg) [inline packages/cosmic-swingset/src/chain-main.js:667]: why is this a safe to assume?

can the data be modeled more robustly?

File: packages/vats/src/localchain.js
Comment (turadg) [inline packages/vats/src/localchain.js:46]: could be helpful to point out in a comment that this relies on the `VLOCALCHAIN_EXECUTE_TX` handler to ensure that the `messages` authenticate with the `address`.

also we should have some tests of that. I don't see them. cc @raphdev

File: scripts/generate-a3p-submission.sh
Comment (turadg) [inline scripts/generate-a3p-submission.sh]: why do you prefer shift to ordinals? I see it's more flexible, but it seems like the positions should be maintained

File: scripts/generate-a3p-submission.sh
Comment (turadg) [inline scripts/generate-a3p-submission.sh]: per points above, I think we should avoid mixing testing logic with the submissions.

**Review by turadg — COMMENTED** [review]
Full pass, requesting some refactoring of the proposal tests.

I didn't review the Golang.

File: packages/vats/src/localchain.js
Comment (turadg) [inline packages/vats/src/localchain.js]: Mind linking here to what those guidelines are? I can't tell whether this does what it claims

File: golang/cosmos/app/app.go
Comment (michaelfig) [reply golang/cosmos/app/app.go]: I was uncertain about whether they should be included in the unreleased upgrade.  I think now that they should.

File: packages/builders/scripts/vats/init-localchain.js
Comment (michaelfig) [reply packages/builders/scripts/vats/init-localchain.js]: Sounds good.  Will do!

File: scripts/generate-a3p-submission.sh
Comment (michaelfig) [reply scripts/generate-a3p-submission.sh]: Vestigial from when I tried allowing additional arguments to the proposals.  Removing.

File: packages/cosmic-swingset/src/chain-main.js
Comment (michaelfig) [reply packages/cosmic-swingset/src/chain-main.js:667]: I'll add a comment in the Golang code that generates the action.

Comment (michaelfig) [reply packages/cosmic-swingset/src/chain-main.js:667]: The data could be modeled more robustly, but there are determinism constraints, such as the inability to JSON-serialize a Golang map consistently.  We'll (cosmic-swingset folks will) tackle those things head-on when we invest some time in split-brain.

File: packages/cosmic-swingset/src/chain-main.js
Comment (michaelfig) [reply packages/cosmic-swingset/src/chain-main.js]: It was cargo-culted from other code.

File: packages/deployment/Makefile
Comment (michaelfig) [reply packages/deployment/Makefile]: I was using it to specify a custom `make docker-build-sdk DOCKER_BUILD="docker build buildx --driver=cloud"` for myself.  I can back out the change.

File: packages/internal/src/config.js
Comment (michaelfig) [reply packages/internal/src/config.js]: I settled on CosmosInitKeyToBridgeId.

File: packages/vats/src/localchain.js
Comment (michaelfig) [reply packages/vats/src/localchain.js:30]: I expect this `allBalances` wrapper to find a better home rather than this expedient one.  It will likely just be moved to a JS test, right now it is just showing an example and a proper Orchestration API will subsume it.

File: packages/vats/src/proposals/localchain-proposal.js
Comment (michaelfig) [reply packages/vats/src/proposals/localchain-proposal.js]: Gotcha.

File: packages/agoric-cli/src/commands/wallet.js
Comment (michaelfig) [reply packages/agoric-cli/src/commands/wallet.js:11]: That was part of getting `agoric run script.js args...` to pass through all args to the script, even if they began with `-`.

Before, commander would interpret top-level options no matter where they occured in the command, and regardless of subcommand.  When I changed this configuration to stop doing that, it required explictly applying those options to the subcommands that wanted them.

The Commander configs to look at are `enablePositionalOptions()` and `passThroughOptions()`.

File: packages/vats/src/localchain.js
Comment (michaelfig) [reply packages/vats/src/localchain.js:46]: I tested this manually, but you're totally right; we rely on this property for security so I've added a test in vlocalchain_test.go.

File: packages/deployment/Makefile
Comment (turadg) [reply packages/deployment/Makefile]: Thanks, I think it's easy enough to tailor locally when needed.

Out of curiosity, what cloud are you using? We have a Depot.dev account now and it can take over buildx without tailoring CLI. `brew install depot && depot configure-docker`

File: a3p-integration/proposals/a:upgrade-next/package.json
Comment (turadg) [inline a3p-integration/proposals/a:upgrade-next/package.json:22]: I suspect this isn't necessary. This proposal package doesn't have a `submission` directory. `b:local-chain` does, and it has a `test-localchain.js` which Ava in most agoric-sdk packages would match because they override Ava's default. But I think the default glob is `*.test.js` so it wouldn't pick up `test-*` here.

Not a blocker because I'm not certain and about to work in this code after you land this so I can clean it up (after verifying).

File: golang/cosmos/app/app.go
Comment (turadg) [inline golang/cosmos/app/app.go]: is there something to fix here? If not I think `XXX` is a more appropriate comment heading

File: golang/cosmos/x/vlocalchain/keeper/keeper.go
Comment (turadg) [inline golang/cosmos/x/vlocalchain/keeper/keeper.go]: consider following the TODO immediately with what is to be done (so greps for "TODO" show it)

File: packages/vats/src/localchain.js
Comment (turadg) [reply packages/vats/src/localchain.js:30]: Got it. Thanks for including the comment on the interface

File: packages/vats/src/proposals/localchain-proposal.js
Comment (turadg) [inline packages/vats/src/proposals/localchain-proposal.js]: I'm cautious about failing silently. Did some CI fail without this?

File: packages/vats/src/proposals/localchain-test.js
Comment (turadg) [inline packages/vats/src/proposals/localchain-test.js]: If `chainStorage` isn't provided, this test succeeds without testing that. 

Please require `chainStorage` so that this can test integration with it. I think it always is where this proposal is being used.

File: packages/vm-config/decentral-demo-config.json
Comment (turadg) [inline packages/vm-config/decentral-demo-config.json]: remove

**Review by turadg — APPROVED** [review]
Thanks for the clear fixup commits. I think you addressed all my comments.

There were a few new things which I've mentioned. I don't see them as blockers because I can address them in master in the next PR. But if you do end up pushing to this branch again, please do handle them.

Again I haven't reviewed the Golang line by line. I did examine the factoring and the test coverage, which LGTM.

File: packages/internal/src/config.js
Comment (turadg) [reply packages/internal/src/config.js]: Yay for explicitness

File: a3p-integration/proposals/a:upgrade-next/package.json
Comment (michaelfig) [reply a3p-integration/proposals/a:upgrade-next/package.json:22]: For whatever reason, when running `b:localchain/test.sh`, without `"!submission"`, `ava` tries to execute `submission/test-localchain.js` as if it were an AVA test, which causes a failure.

You're right that it isn't strictly necessary for `a:` and `g:`, but I thought it was important to keep these `package.json` files in sync, since people may copy them and use them as the basis for their own proposal test.

So I'll add it for now and leave it to you to clarify as you see fit.

File: golang/cosmos/app/app.go
Comment (michaelfig) [reply golang/cosmos/app/app.go]: I'll make it a `NOTE`.  Maybe you can explain what `XXX` means if you think it's better?

File: packages/vats/src/proposals/localchain-proposal.js
Comment (michaelfig) [reply packages/vats/src/proposals/localchain-proposal.js]: I'm accustomed to working in the core behaviours of `@agoric/vats` where code is loaded unconditionally, rather in these core-evals that will never run off-chain.

So you're right, I've made this code throw an error if it's missing a bridge.

File: packages/vm-config/decentral-demo-config.json
Comment (michaelfig) [reply packages/vm-config/decentral-demo-config.json]: I've run `test-localchain.js` explicitly now.

File: packages/deployment/Makefile
Comment (michaelfig) [reply packages/deployment/Makefile]: Oh, at the time I was playing with Docker buildcloud, just for funsies.

File: golang/cosmos/app/app.go
Comment (turadg) [reply golang/cosmos/app/app.go]: sure (I should have included this ref with my last comment): https://github.com/Agoric/agoric-sdk/wiki/Coding-Style#code-comments

File: packages/vats/src/proposals/localchain-proposal.js
Comment (turadg) [reply packages/vats/src/proposals/localchain-proposal.js]: makes sense

File: a3p-integration/proposals/a:upgrade-next/package.json
Comment (turadg) [reply a3p-integration/proposals/a:upgrade-next/package.json:22]: Ah, it must be inheriting the config from the SDK root package.json. I will address later

File: packages/vats/src/proposals/localchain-proposal.js
Comment (michaelfig) [reply packages/vats/src/proposals/localchain-proposal.js]: Erk.  Unfortunately, the sim-chain configuration of `ag-solo` uses the same config as the real chain, but lacks a bridgeManager.  I'm reverting the throw.

```
Run cd packages/solo && yarn test | $TEST_COLLECT
yarn run v1.22.19
$ ava
TAP version 13
# connecting.........................Loading slog sender modules: @agoric/telemetry/src/flight-recorder.js
................Loading slog sender modules: @agoric/telemetry/src/flight-recorder.js
.oo2024-02-21T21:[15](https://github.com/Agoric/agoric-sdk/actions/runs/7995614069/job/21836438318#step:5:16):49.920Z SwingSet: xsnap: v1: Error#1 ERROR_NOTE: Sent as error:liveSlots:v1#70001
2024-02-21T21:15:49.921Z SwingSet: xsnap: v1: Error#1 ERROR_NOTE: Sent as error:liveSlots:v1#70002
2024-02-21T21:15:49.922Z SwingSet: kernel: ##### KERNEL PANIC: kp40.policy panic: rejected {"body":"#{\"#error\":\"No bridgeManager, cannot setupLocalChainVat\",\"errorId\":\"error:liveSlots:v1#70002\",\"name\":\"Error\"}","slots":[]} #####
2024-02-21T21:15:49.936Z SwingSet: kernel: ##### KERNEL PANIC: kp40.policy panic: rejected {"body":"#{\"#error\":\"vat terminated\",\"name\":\"Error\"}","slots":[]} #####
2024-02-21T21:15:49.936Z SwingSet: kernel: ##### KERNEL PANIC: critical vat v1 failed #####
[20](https://github.com/Agoric/agoric-sdk/actions/runs/7995614069/job/21836438318#step:5:21)24-02-[21](https://github.com/Agoric/agoric-sdk/actions/runs/7995614069/job/21836438318#step:5:22)T21:15:49.937Z block-manager: AG_COSMOS_INIT error: (Error#1)
Error#1: #{"#error":"No bridgeManager, cannot setupLocalChainVat","errorId":"error:liveSlots:v1#70001","name":"Error"}
```

Comment (turadg) [reply packages/vats/src/proposals/localchain-proposal.js]: Gotcha. A comment as to why would help future readers

File: golang/cosmos/x/vlocalchain/keeper/keeper.go
Comment (raphdev) [inline golang/cosmos/x/vlocalchain/keeper/keeper.go:180]: Is this another case of "signer" being misleading? AFAICT there are no signatures involved.

File: golang/cosmos/x/vlocalchain/keeper/keeper.go
Comment (raphdev) [inline golang/cosmos/x/vlocalchain/keeper/keeper.go:209]: Is it typical to allow bundling multiple messages, and succeed only if all messages succeed? Is it obvious to smart contracts that unrelated messages stuffed in a single tx may fail if an unrelated message fails?

**Review by raphdev — APPROVED** [review]
I reviewed the golang code closer than I did before. LGTM. Just a few questions for my understanding that are unlikely to be issues.

File: golang/cosmos/x/vlocalchain/keeper/keeper.go
Comment (raphdev) [reply golang/cosmos/x/vlocalchain/keeper/keeper.go:209]: Confirmed this is the case.

File: golang/cosmos/x/vlocalchain/types/key.go
Comment (JimLarson) [inline golang/cosmos/x/vlocalchain/types/key.go]: Note that it stores a big-endian uint64. Consider whether to use a proto as the value so we can add new fields in the future.

File: golang/cosmos/x/vlocalchain/types/key.go
Comment (JimLarson) [inline golang/cosmos/x/vlocalchain/types/key.go]:
```suggestion
// IncrementSequence increments its argument as a big-endian Uint64.
func IncrementSequence(value []byte) []byte {
```

File: golang/cosmos/x/vlocalchain/keeper/keeper.go
Comment (JimLarson) [inline golang/cosmos/x/vlocalchain/keeper/keeper.go]: Docstring and add tests, here and for `serviceToTypeUrl()` too.

Consider unifying into:
```
fullMethod, responseURL := parseRequestTypeUrl(anyMsg.TypeUrl)
```

File: golang/cosmos/proto/agoric/vlocalchain/vlocalchain.proto
Comment (JimLarson) [inline golang/cosmos/proto/agoric/vlocalchain/vlocalchain.proto]: formatting

File: golang/cosmos/x/vlocalchain/keeper/keeper.go
Comment (JimLarson) [inline golang/cosmos/x/vlocalchain/keeper/keeper.go]: `ValidateBasic` validation is independent of chain state, right? Consider validating all messages before starting the more expensive execution.

File: golang/cosmos/x/vlocalchain/keeper/keeper.go
Comment (JimLarson) [inline golang/cosmos/x/vlocalchain/keeper/keeper.go:250]: Panic if `len(res.MsgResponses) != 1` rather than silently ignoring extra responses.

**Review by JimLarson — APPROVED** [review]
Approving the golang code, with some minor/optional comments I trust you to address.

**Comment by schnetzlerjoe** [comment]
@michaelfig This is a massive improvement on dev x I just want to add. It will reduce our Calypso code base substantially.

**Pull request merged**
